text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,206 @@
+inverse is not allowed. Not only does this make things easier to reason about, but many existing
+components rely on workarounds in synthetic mode that are not possible to support in native mode.
+These observable differences are further discussed below. As this invariant cannot be asserted
+during compile-time, the engine will assert it during runtime.
+
+### Testing
+
+LWC currently has integration tests for synthetic shadow mode and native shadow mode. A third
+testing mode called "mixed shadow mode" will be introduced. Synthetic shadow DOM polyfills are
+implemented as patches on global prototypes so the decision of invoking a polyfilled API vs a native
+API will depend on the preference of individual components during runtime. This test environment
+will ensure that components designed to run in a native Shadow DOM environment will work correctly
+when even when synthetic Shadow DOM polyfills are present.
+
+Components that prefer native mode will need to also support synthetic mode in browser environments
+where Shadow DOM is natively unavailable, so they will also need to be tested in both modes. LWC
+test utilities will need to be updated to facilitate this.
+
+The existing WPT (Web Platform Tests) test suite for Shadow DOM APIs will also be run to identify
+any coverage gaps in LWC integration tests.
+
+## Observable differences
+
+### Light DOM and assigned elements
+
+In native Shadow DOM, elements are rendered in a component's light DOM and assigned to slots in a
+child component's shadow DOM. This means that elements will exist in the DOM regardless of whether
+they are assigned to a slot. This is not the case for LWC--elements that are passed down to child
+components are never rendered unless they are assigned to a slot.
+
+In the following example, `span` will not exist in the DOM in synthetic shadow, but it will exist in
+native shadow.
+
+```html
+<!-- x-parent -->
+<template>
+    <x-child>
+        <span>foo</span>
+    </x-child>
+</template>
+
+<!-- x-child -->
+<template>
+    <p>child</p>
+</template>
+```
+
+This observable difference will not be remediated as such a change would be non-trivial and
+backwards-incompatible with the current synthetic shadow implementation.
+
+### Lifecycle timing
+
+Due to the way that slotting is implemented in `@lwc/synthetic-shadow`, there is an observable
+difference in the timing of lifecycle hooks for slotted elements.
+
+As mentioned previously, in synthetic mode, slotted elements that are never assigned to a slot are
+not rendered. This means that their lifecycle hooks are never invoked.
+
+Another difference is that for synthetic mode, lifecycle hooks are invoked in the order of
+appearance after they are assigned, whereas in native mode, they are invoked in the order of
+appearance in the template.
+
+### this.shadowRoot vs this.template
+
+`this.shadowRoot` and `this.template` will both reference the component's shadow root in native

```

Comment:
I disagree with this as a signal because this will put in danger any attempt to merge component and elements into a single instance in the future because the getter for shadowRoot is subject to the mode used for the shadow, which is orthogonal to this proposal.

Reply:
I'm not sure I understand. Are you saying that we should avoid exposing `this.shadowRoot` because this is something we will never do for LWC components?

Reply:
No, I'm saying that if we ever merge this instance with the element instance, it will be the element instance the one deciding whether or not that is null based on the attach shadow mode.

Reply:
It looks like `this.shadowRoot` is also not settable, even for closed-mode shadow elements. So we can't override the default behavior if we merge the two classes.

![Screen Shot 2021-06-03 at 3 58 52 PM](https://user-images.githubusercontent.com/283842/120722459-9df8fd80-c484-11eb-87f0-8dee4ba381fb.png)

Although I suppose even in the case where we merge the two classes and support closed shadow mode, we could just allow `shadowRoot` to be null? Would this be a problem?

Reply:
@ekashida and I chatted about this. Since exposing `this.shadowRoot` would introduce new security concerns in the SA (since `this.template` is allowlisted, whereas `this.shadowRoot` allows unfiltered access), and since it's orthogonal to the mixed mode feature, we're going to punt on this for now. We can always consider how to expose `this.shadowRoot` later.